### ISSUE TITLE
dl.elv.sh unavailable (SSL error)

### ISSUE THREAD

**Issue by fgimian** [elves/elvish#1614]
Hey there, I cannot download Elvish via Scoop or using curl at the moment:

```
❯ curl -O -v https://dl.elv.sh/windows-amd64/elvish-v0.18.0.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 172.93.189.248:443...
* Connected to dl.elv.sh (172.93.189.248) port 443 (#0)
* schannel: disabled automatic use of client certificate
* ALPN: offers http/1.1
* schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Closing connection 0
curl: (35) schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.

~
```

This appears to be a similar issue to #1589 and #1558.  Can you assist please?

Thanks
Fotis

**Comment by krader1961** [comment]
Note that 172.93.189.248 is a reserved, private, address. It is not the address of dl.elv.sh. There is something wrong with your (corporate?) firewall or proxy.

**Comment by fgimian** [comment]
> Note that 172.93.189.248 is a reserved, private, address. It is not the address of dl.elv.sh. There is something wrong with your (corporate?) firewall or proxy.

Thanks for the reply.  I'm actually on a direct connection and can replicate this IP being returned via both Cloudflare and Google DNS servers:

```
🕙 [ 11:15:00 PM ] ❯ nslookup.exe dl.elv.sh 1.1.1.1
Server:  one.one.one.one
Address:  1.1.1.1

Non-authoritative answer:
Name:    dl.elv.sh
Address:  172.93.189.248


~
🕙 [ 11:15:04 PM ] ❯ nslookup.exe dl.elv.sh 8.8.8.8
Server:  dns.google
Address:  8.8.8.8

Non-authoritative answer:
Name:    dl.elv.sh
Address:  172.93.189.248
```

In the meantime, I installed elvish with `go install` 😄 

Cheers
Fotis

**Comment by fgimian** [comment]
The IP address being resolved appears to be a Debian server:

```
❯ telnet 172.93.189.248 22
Trying 172.93.189.248...
Connected to 172.93.189.248.
Escape character is '^]'.
SSH-2.0-OpenSSH_8.4p1 Debian-5
↔q

telnet.exe"> q
Connection closed.
```

**Comment by fgimian** [comment]
This website shows the issue clearly: https://www.whatsmydns.net/#A/elv.sh

The internal address is being returned for the Asia Pacific region including Australia and seemingly Asia too.

**Comment by krader1961** [comment]
@fgimian: Yeah, you're right. My bad. Whenever I see someone open an issue like this with a 10., 172., or 192.168. address I assume it's a firewall/proxy problem because that is almost always the case. However, I see, on closer examination, the address you're getting is outside the 172.16. to 172.31. private address range. Hopefully @xiaq notices this issue soon and fixes whatever is wrong with either the DNS configuration for the project or that specific server.

**Comment by krader1961** [comment]
I took another look. According to https://dnschecker.org/#A/dl.elv.sh host `dl.elv.sh` is still resolving to IP addr 172.93.189.248 for sites originating in the Asia-Pacific region. Attempting to download from it still fails:

```
 macpro  ~  > curl -O -v https://172.93.189.248/windows-amd64/elvish-v0.18.0.zip              7.7 m
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 172.93.189.248:443...
* Connected to 172.93.189.248 (172.93.189.248) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
} [296 bytes data]
* error:1404B438:SSL routines:ST_CONNECT:tlsv1 alert internal error
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Closing connection 0
curl: (35) error:1404B438:SSL routines:ST_CONNECT:tlsv1 alert internal error
```

@xiaq: IP address 172.93.189.248 is apparently hosted by Gigabit Hosting Sdn Bhd based in Malaysia. Perhaps that server needs a reboot and/or reconfiguration. If that doesn't fix the problem perhaps that server should be removed from the DNS config for `dl.elv.sh`. it is obviously preferable for downloads by people in the AP region to be slow rather than fail.

**Comment by iandol** [comment]
FWIW: I work in China and dl.elv.sh haasn't worked for ages, and I use the mirror: https://mirrors.tuna.tsinghua.edu.cn/elvish

**Comment by krader1961** [comment]
@iandol: Good to know. I worked in the AP region for 19 months from 1997 to 1999. So this issue doesn't affect me today but I'm still cognizant that people outside of the USA should be supported as well as those inside the USA. Even though I'm an American I try my best not to be provincial with respect to where I was born. 😄

**Comment by xiaq** [comment]
Sorry I missed this issue.

I am not able to reproduce the issue from my machine in APAC.

@fgimian Are you able to access any of the following?

- https://cdg.dl.elv.sh
- https://hkg.dl.elv.sh

dl.elv.sh indeed resolves to different hosts depending on the region (the two links above are directly access to the backend hosts with no geo DNS). If you can't access either of the above, it's likely because of some issue with the HTTPS client.

I've seen issues where the client uses an old insecure cipher and I've configured my servers to reject those. That may not be the only cause though.

FWIW, using HTTPS with a direct IP is unlikely to be useful for debugging because HTTPS certificates are usually tied to domains.

**Comment by xiaq** [comment]
This tool thinks that connecting to https://dl.elv.sh is fine globally: https://latency.apex.sh/?url=https%3A%2F%2Fdl.elv.sh%2F&compare=

**Comment by krader1961** [comment]
@xiaq: If you check with https://dnschecker.org/#A/dl.elv.sh you'll see that it still resolves to 172.93.189.248 for some locations in the Asia-Pacific region like Kuala Lumpur. It seems likely that latency.apex.sh is resolving to a different address (probably 51.158.65.224).

> FWIW, using HTTPS with a direct IP is unlikely to be useful for debugging because HTTPS certificates are usually tied to domains.

Yeah, I was simply trying to see if there was a working HTTPS server at that address and who owned it.

**Comment by fgimian** [comment]
It does seem to be working correctly here now, even though it's still resolving to the 172 address.  So I think this issue is solved.  Feel free to close it if you're satisfied that it's working too. 😄 

Thanks!
Fotis

**Comment by xiaq** [comment]
Hmm, I did update the version of Caddy at some point. So maybe it was Caddy that fixed a compatibility issue with your client. Anyway, thanks for checking - I'll close this now.

**Comment by rktjmp** [comment]
Seems like something might be wonky here (again?).

`elv.sh` resolves to `172.93.189.248` for me from Australia (home internet, mobile internet & a VM in Austraila), as well as from vms in Japan & Singapore.

```
λ drill elv.sh @1.1.1.1
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 40891
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; QUESTION SECTION:
;; elv.sh.      IN      A

;; ANSWER SECTION:
elv.sh. 31      IN      A       172.93.189.248

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 28 msec
;; SERVER: 1.1.1.1
;; WHEN: Wed Aug 13 20:34:24 2025
;; MSG SIZE  rcvd: 40
```

[NSLookup.io](https://www.nslookup.io/domains/elv.sh/dns-propagation/a/) results can vary but consistently shows at least two variants, generally split by Asia-Pacific & Rest-of-world. Note that it reliably shows google, cloudflare and quad9s returning the "wrong variant", it's not just some minor platform/isp with a misconfigured dns server.

<img width="750" height="615" alt="Image" src="https://github.com/user-attachments/assets/bc2b1fa7-9cd8-48b4-a2e9-2ff760f8454f" />

<img width="774" height="426" alt="Image" src="https://github.com/user-attachments/assets/1c113259-7fae-4c50-bf2d-87da3bfd6138" />



https://www.uptrends.com/tools/uptime

<img width="829" height="631" alt="Image" src="https://github.com/user-attachments/assets/e636659f-898a-45ca-830a-23f52ce8a383" />

https://check-host.net/check-http?host=elv.sh

<img width="807" height="785" alt="Image" src="https://github.com/user-attachments/assets/7262fb78-fdae-4d44-aba6-73a6912475a2" />

Note that many of the non-functioning regions seem to be missing an AAAA record, which may indicate a difference in configuration for those regions?

https://check-host.net/check-dns?host=elv.sh

<img width="817" height="754" alt="Image" src="https://github.com/user-attachments/assets/2852c13e-b68c-466f-ab87-aa7e13869c36" />

**Comment by xiaq** [comment]
@rktjmp Yeah, there's a separate host in Hong Kong for Asia-Pacific IPs because the connectivity between the main host in France and Asia-Pacific is pretty bad.

Unfortunately the HK hosting provider is not the most reliable. I've disabled DNS resolution to that host for now, connectivity from Asia Pacific should be resumed when the record propagates.

I'll pick another provider and add some proper monitoring before re-enabling this setup.
